oauth-keycloak-authorizer/src/main/java/io/strimzi/kafka/oauth/server/authorizer/GrantsHandler.java

```diff
@@ -0,0 +1,520 @@
+                scheduled.add(new Future(userId, grantsInfo, refreshWorker.submit(() -> {
+
+                    if (log.isTraceEnabled()) {
+                        log.trace("Fetch grants for user: " + userId + ", token: " + mask(grantsInfo.getAccessToken()));
+                    }
+
+                    JsonNode newGrants;
+                    try {
+                        newGrants = fetchGrantsWithRetry(grantsInfo.getAccessToken());
+                    } catch (HttpException e) {
+                        if (403 == e.getStatus()) {
+                            // 403 happens when no policy matches the token - thus there are no grants, no permission granted
+                            newGrants = JSONUtil.newObjectNode();
+                        } else {
+                            throw e;
+                        }
+                    }
+                    JsonNode oldGrants = grantsInfo.getGrants();
+                    if (!semanticGrantsEquals(newGrants, oldGrants)) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Grants have changed for user: {}; before: {}; after: {}", userId, oldGrants, newGrants);
+                        }
+                        grantsInfo.setGrants(newGrants);
+                    }
+
+                    // Only added here to allow compiler to resolve the lambda as a Callable<JsonNode>
+                    return newGrants;
+                })));
+            }
+
+            for (GrantsHandler.Future f : scheduled) {
+                try {
+                    f.get();
+                } catch (ExecutionException e) {
+                    final Throwable cause = e.getCause();
+                    if (cause instanceof HttpException) {
+                        log.debug("[IGNORED] Failed to fetch grants for user: {}", cause.getMessage());
+                        if (401 == ((HttpException) cause).getStatus()) {
+                            grantsCache.remove(f.getUserId());
+                            log.debug("Removed user from grants cache: {}", f.getUserId());
+                            Services.getInstance().getSessions().removeAllWithMatchingAccessToken(f.getGrantsInfo().accessToken);
+                            continue;
+                        }
+                    }
+                    if (log.isWarnEnabled()) {
+                        log.warn("[IGNORED] Failed to fetch grants for user: " + e.getMessage(), e);
+                    }
+
+                } catch (Throwable e) {
+                    if (log.isWarnEnabled()) {
+                        log.warn("[IGNORED] Failed to fetch grants for user: " + f.getUserId() + ", token: " + mask(f.getGrantsInfo().accessToken) + " - " + e.getMessage(), e);
+                    }
+                }
+            }
+
+        } catch (Throwable t) {
+            // Log, but don't rethrow the exception to prevent scheduler cancelling the scheduled job.
+            log.error(t.getMessage(), t);
```

Comment:
If `t.getMessage()` contained `{}` then message that was logged would be different than if `t.getMessage()` didn't contain `{}`. So this should probably be `log.error("{}", t.getMessage(), t);` ?

Reply:
Whoa, you just blew my mind here! You are absolutely right! I never understood this properly, and it went by me so easy.
There should be a Spotbugs rule for this.